src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombiner.java

```diff
@@ -61,34 +96,177 @@ private void combineShardScores(final ScoreCombinationTechnique scoreCombination
 
         // - sort documents by scores and take first "max number" of docs
         // create a collection of doc ids that are sorted by their combined scores
-        List<Integer> sortedDocsIds = getSortedDocIds(combinedNormalizedScoresByDocId);
+        Collection<Integer> sortedDocsIds = getSortedDocIds(
+            combinedNormalizedScoresByDocId,
+            getTopFieldDocs(sort, topDocsPerSubQuery),
+            sort
+        );
 
         // - update query search results with normalized scores
-        updateQueryTopDocsWithCombinedScores(compoundQueryTopDocs, topDocsPerSubQuery, combinedNormalizedScoresByDocId, sortedDocsIds);
+        updateQueryTopDocsWithCombinedScores(
+            compoundQueryTopDocs,
+            topDocsPerSubQuery,
+            combinedNormalizedScoresByDocId,
+            sortedDocsIds,
+            getDocIdSortFieldsMap(compoundQueryTopDocs, combinedNormalizedScoresByDocId, sort),
+            sort != null
+        );
     }
 
-    private List<Integer> getSortedDocIds(final Map<Integer, Float> combinedNormalizedScoresByDocId) {
+    private boolean checkIfSortOrderByScore(Sort sort) {
+        if (sort == null) {
+            return false;
+        }
+
+        for (SortField sortField : sort.getSort()) {
+            if (sortField.getType().equals(SortField.Type.SCORE)) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     *
+     * @param sort sort criteria
+     * @param topDocsPerSubQuery top docs per subquery
+     * @return list of top field docs which is deduced by typcasting top docs to top field docs.
+     */
+    private List<TopFieldDocs> getTopFieldDocs(final Sort sort, final List<TopDocs> topDocsPerSubQuery) {
+        if (sort == null) {
+            return null;
+        }
+        List<TopFieldDocs> topFieldDocs = new ArrayList<>();
+        for (TopDocs topDocs : topDocsPerSubQuery) {
+            // Check for scoreDocs length.
+            // If scoreDocs length=0 then it means that no results are found for that particular subquery.
+            if (topDocs.scoreDocs.length != 0) {
```

Comment:
can topDocs.scoreDocs be null?

Reply:
Nope. https://github.com/opensearch-project/neural-search/blob/25d2e82b62b52eae4d73f8f33aab704ff01c0390/src/main/java/org/opensearch/neuralsearch/processor/CompoundTopDocs.java#L78